File: framework/web-core/src/main/java/top/bettercode/simpleframework/servlet/HandlerMethodContextHolder.java
package top.bettercode.simpleframework.servlet;

import java.lang.annotation.Annotation;
import javax.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.method.HandlerMethod;
import org.springframework.web.servlet.HandlerExecutionChain;
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping;
import top.bettercode.simpleframework.AnnotatedUtils;

/**
 * @author Peter Wu
 */
public class HandlerMethodContextHolder {

  private static final Logger log = LoggerFactory.getLogger(HandlerMethodContextHolder.class);
  private static final String HANDLER_METHOD =
      HandlerMethodContextHolder.class.getName() + ".handlerMethod";
  private static RequestMappingHandlerMapping handlerMapping;


  public static void setHandlerMapping(
      RequestMappingHandlerMapping handlerMapping) {
    HandlerMethodContextHolder.handlerMapping = handlerMapping;
  }

  public static HandlerMethod getHandler(HttpServletRequest request) {
    try {
      HandlerMethod handlerMethod = (HandlerMethod) request.getAttribute(
          HandlerMethodContextHolder.HANDLER_METHOD);
      if (handlerMethod != null) {
        return handlerMethod;
      }
      HandlerExecutionChain handlerExecutionChain = handlerMapping.getHandler(request);
      if (handlerExecutionChain != null) {
        Object handler = handlerExecutionChain.getHandler();
        if (handler instanceof HandlerMethod) {
          handlerMethod = (HandlerMethod) handler;
          request.setAttribute(HandlerMethodContextHolder.HANDLER_METHOD, handlerMethod);
          return handlerMethod;
        }
      }
    } catch (Exception e) {
      log.warn(e.getMessage(), e);
    }
    return null;
  }

  public static <A extends Annotation> boolean hasAnnotation(HttpServletRequest request,
      Class<A> annotationType) {
    HandlerMethod handler = getHandler(request);
    return handler != null && AnnotatedUtils.hasAnnotation(handler, annotationType);
  }

  public static <A extends Annotation> A getAnnotation(HttpServletRequest request,
      Class<A> annotationType) {
    HandlerMethod handler = getHandler(request);
    return handler == null ? null : AnnotatedUtils.getAnnotation(handler, annotationType);
  }
}

File: web/src/main/kotlin/top/bettercode/summer/web/servlet/HandlerMethodContextHolder.kt
package top.bettercode.summer.web.servlet

import org.slf4j.LoggerFactory
import org.springframework.boot.web.servlet.error.ErrorController
import org.springframework.http.HttpMethod
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.HandlerMapping
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping
import top.bettercode.summer.tools.lang.operation.HttpOperation
import top.bettercode.summer.tools.lang.operation.RequestConverter
import top.bettercode.summer.tools.lang.util.AnnotatedUtils.getAnnotation
import top.bettercode.summer.tools.lang.util.AnnotatedUtils.hasAnnotation
import javax.servlet.http.HttpServletRequest

/**
 * @author Peter Wu
 */
object HandlerMethodContextHolder {
    private val log = LoggerFactory.getLogger(HandlerMethodContextHolder::class.java)
    private val HANDLER_METHOD = HandlerMethodContextHolder::class.java.name + ".handlerMethod"
    private var handlerMapping: RequestMappingHandlerMapping? = null
    fun setHandlerMapping(
        handlerMapping: RequestMappingHandlerMapping?
    ) {
        HandlerMethodContextHolder.handlerMapping = handlerMapping
    }

    @JvmStatic
    fun getHandler(request: HttpServletRequest): HandlerMethod? {
        try {
            var handler = request.getAttribute(HandlerMapping.BEST_MATCHING_HANDLER_ATTRIBUTE)
            if (handler is HandlerMethod && !ErrorController::class.java.isAssignableFrom(
                    handler.beanType
                )
            ) {
                return handler
            }
            var handlerMethod: HandlerMethod?
            handlerMethod = request.getAttribute(HANDLER_METHOD) as HandlerMethod?
            if (handlerMethod != null) {
                return handlerMethod
            }
//            if (!ServletRequestPathUtils.hasParsedRequestPath(request)) {
//                ServletRequestPathUtils.parseAndCache(request)
//            }
            val handlerExecutionChain = handlerMapping!!.getHandler(request)
            if (handlerExecutionChain != null) {
                handler = handlerExecutionChain.handler
                if (handler is HandlerMethod && !ErrorController::class.java.isAssignableFrom(
                        handler.beanType
                    )
                ) {
                    handlerMethod = handler
                    request.setAttribute(HANDLER_METHOD, handlerMethod)
                    return handlerMethod
                }
            }
        } catch (e: Exception) {
            if (HttpMethod.OPTIONS.name != request.method) {
                val requestString = HttpOperation.toString(
                    request = RequestConverter.convert(request)
                )
                log.warn("\n$requestString", e)
            }
        }
        return null
    }

    fun <A : Annotation?> hasAnnotation(
        request: HttpServletRequest,
        annotationType: Class<A>
    ): Boolean {
        val handler = getHandler(request)
        return handler != null && hasAnnotation(handler, annotationType)
    }

    fun <A : Annotation?> getAnnotation(
        request: HttpServletRequest,
        annotationType: Class<A>
    ): A? {
        val handler = getHandler(request)
        return if (handler == null) null else getAnnotation(handler, annotationType)
    }
}